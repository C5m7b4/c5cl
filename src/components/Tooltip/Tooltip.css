.mcl-tooltip {
  position: relative;
}

.mcl-tooltip-wrapper {
  position: relative;
  min-width: 120px;
  min-height: 120px;
}

.mcl-tooltip-trigger {
  display: inline-block;
}

.mcl-tooltip-bubble {
  min-width: 120px;
  position: absolute;
  z-index: 10;
  transition: opacity 750ms;
}

.mcl-tooltip-bubble::after {
  content: '';
  position: absolute;
}

.mcl-tooltip-top {
  bottom: 100%;
  left: 50%;
  padding-bottom: 9px;
  transform: translateX(-50%);
}

.mcl-tooltip-top::after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-top: 9px solid #fff;
  bottom: 0;
  left: 50%;
  transform: translateX(-50%);
}

.mcl-tooltip-bottom {
  top: 100%;
  left: 50%;
  padding-top: 9px;
  transform: translateX(-50%);
}

.mcl-tooltip-bottom::after {
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  border-bottom: 9px solid #fff;
  top: 0;
  left: 50%;
  transform: translateX(-50%);
}

.mcl-tooltip-left {
  top: 50%;
  right: 100%;
  padding-right: 9px;
  transform: translateY(-50%);
}

.mcl-tooltip-left::after {
  border-left: 9px solid #fff;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  right: 0;
  transform: translateY(-50%);
}

.mcl-tooltip-right {
  top: 50%;
  left: 100%;
  padding-left: 9px;
  transform: translateY(-50%);
}

.mcl-tooltip-right::after {
  border-right: 9px solid #fff;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 50%;
  left: 0;
  transform: translateY(-50%);
}

.mcl-tooltip-message {
  background: #fff;
  border-radius: 3px;
  color: #000;
  font-size: 0.75rem;
  z-index: 9999;
}
