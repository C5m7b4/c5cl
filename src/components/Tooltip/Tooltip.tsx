import React, { useState, useRef, useEffect } from 'react';

import './Tooltip.css';

export type TooltipPosition = 'top' | 'bottom' | 'left' | 'right';

export type TooltipTheme = 'light' | 'dark';

export interface TooltipProps {
  position: TooltipPosition;
  children: React.ReactNode;
  message: string;
  style?: React.CSSProperties;
  theme?: TooltipTheme;
  messageStyle?: React.CSSProperties;
}

const Tooltip = (props: TooltipProps) => {
  const [displayTooltip, setDisplayTooltip] = useState(true);
  const { position = 'top' } = props;

  let internalStyle = {
    backgroundColor: '#fff',
    color: '#000',
    padding: '5px',
  } as React.CSSProperties;

  if (props.theme === 'dark') {
    internalStyle = {
      backgroundColor: '#000',
      color: '#fff',
      padding: '3px',
    } as React.CSSProperties;
  }

  const tooltipRef = useRef<HTMLSpanElement>(null);
  const triggerRef = useRef<HTMLSpanElement>(null);

  useEffect(() => {
    /* istanbul ignore else */
    if (props.style) {
      internalStyle = props.style;
    }
  }, []);

  const showTooltip = () => {
    setDisplayTooltip(true);
  };

  const hideTooltip = () => {
    setDisplayTooltip(false);
  };

  return (
    <span
      className="mcl-tooltip-wrapper"
      ref={tooltipRef}
      onMouseEnter={showTooltip}
      onMouseLeave={hideTooltip}
    >
      {displayTooltip ? (
        <div
          className={`mcl-tooltip-bubble mcl-tooltip-${position} ${
            props.theme === 'dark' ? 'dark' : ''
          }`}
        >
          <div
            className="mcl-tooltip-message"
            style={{ ...internalStyle, ...props.messageStyle }}
          >
            {props.message}
          </div>
        </div>
      ) : null}
      <span ref={triggerRef} className="mcl-tooltip-trigger">
        {props.children}
      </span>
    </span>
  );
};

export default Tooltip;
