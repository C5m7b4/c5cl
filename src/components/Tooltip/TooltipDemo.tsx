import React from 'react';

import Tooltip from './Tooltip';

import { TooltipPosition, TooltipTheme } from './Tooltip';

export interface TooltipDemoProps {
  position: TooltipPosition;
  message: string;
  theme: TooltipTheme;
  style?: React.CSSProperties;
  messageStyle?: React.CSSProperties;
}

const TooltipDemo = (props: TooltipDemoProps) => {
  return (
    <div>
      <Tooltip
        position={props.position}
        message={props.message}
        style={{ marginTop: '20px' }}
        theme={props.theme}
        messageStyle={props.messageStyle}
      >
        <button>{props.message}</button>
      </Tooltip>
    </div>
  );
};

export default TooltipDemo;
