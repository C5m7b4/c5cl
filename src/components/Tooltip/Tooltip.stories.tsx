import React from 'react';
import { Story, Meta } from '@storybook/react';

import TooltipDemo, { TooltipDemoProps } from './TooltipDemo';

export default {
  title: 'Widgets/Tooltip',
  component: TooltipDemo,
  parameters: {
    layout: 'centered',
  },
} as Meta;

const Template: Story<TooltipDemoProps> = (args) => <TooltipDemo {...args} />;

export const TooltipDemoExampleTop = Template.bind({});
TooltipDemoExampleTop.args = {
  position: 'top',
  message: 'Top Tooltip',
  theme: 'light',
  messageStyle: {
    padding: '20px',
  },
};

export const TooltipDemoExampleBottom = Template.bind({});
TooltipDemoExampleBottom.args = {
  position: 'bottom',
  message: 'Bottom Tooltip',
};

export const TooltipDemoExampleLeft = Template.bind({});
TooltipDemoExampleLeft.args = {
  position: 'left',
  message: 'Left Tooltip',
};

export const TooltipDemoExampleRight = Template.bind({});
TooltipDemoExampleRight.args = {
  position: 'right',
  message: 'Right Tooltip',
};
