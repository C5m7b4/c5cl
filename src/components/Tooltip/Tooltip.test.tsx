import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import Tooltip from './Tooltip';
import { act } from 'react-dom/test-utils';

describe('Tooltip', () => {
  test('should render correctly', () => {
    const { container } = render(
      <Tooltip position="top" message="tooltip">
        <h2>Hello</h2>
      </Tooltip>
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle some style', () => {
    const { container } = render(
      <Tooltip
        position="top"
        message="tooltip"
        style={{ backgroundColor: 'red' }}
      >
        <h2>Hello</h2>
      </Tooltip>
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle mouseOver', () => {
    const { container, getByRole, unmount } = render(
      <Tooltip position="top" message="tooltip">
        <div>Hello</div>
      </Tooltip>
    );
    const header = container.querySelector(
      '.mcl-tooltip-trigger'
    ) as HTMLDivElement;
    fireEvent.mouseOver(header);
    fireEvent.mouseLeave(header);
    fireEvent.mouseOut(header);
  });
  test('should run useEffect cleanup return function', () => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
    const { unmount } = render(
      <Tooltip position="top" message="tooltip">
        <div>Hello</div>
      </Tooltip>
    );
    unmount();
  });
});
