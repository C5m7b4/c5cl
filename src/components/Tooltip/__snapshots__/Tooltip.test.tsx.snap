// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should handle some style 1`] = `
<div>
  <span
    class="mcl-tooltip-wrapper"
  >
    <span
      class="mcl-tooltip-trigger"
    >
      <h2>
        Hello
      </h2>
    </span>
  </span>
</div>
`;

exports[`Tooltip should render correctly 1`] = `
<div>
  <span
    class="mcl-tooltip-wrapper"
  >
    <span
      class="mcl-tooltip-trigger"
    >
      <h2>
        Hello
      </h2>
    </span>
  </span>
</div>
`;
